### PR TITLE
Add support for more audience fields.

### PR DESCRIPTION
**Pull request by aontas**
This adds first class support for _last_used and times_used audience fields, instead of just ignoring them.

### CODE REVIEW

**Comment by natanrolnik** [comment]
Thanks for your PR, @aontas 

It seems that the test `should update audiences` is failing:

```
    ✗ should update audiences
      - Failed: Cannot read property 'get' of undefined
```

Are you able to make sure that tests pass locally?

**Comment by aontas** [comment]
I had, though running them multiple times leaves artefacts which makes what would have been undefined, defined. I'll fix the issue and try again.

**Comment by natanrolnik** [comment]
This is very weird. Code coverage decreasing bt 54% 😲

File: spec/PushController.spec.js [hunk 1]
```diff
@@ -979,4 +979,93 @@ describe('PushController', () => {
       done();
     }).catch(done.fail);
   });
+
+  fit('should update audiences', (done) => {
```
Comment (flovilmart) [inline spec/PushController.spec.js]: There... that’s the issue. Should be ‘it’

**Comment by aontas** [comment]
I was testing the tests on Travis and only had 1 test forced to run. The tends to reduce the coverage a large amount.

@flovilmart can you add your comment again? I rebased and I never had the opportunity to read your comment.

**Comment by aontas** [comment]
Has the right fields for dashboard support (hence the mongo transform updates).

**Comment by flovilmart** [comment]
@aontas there was the PushAudienceId that could be stored alongside, so probably, we should set that from the dashboard instead of the inference based upon the query don't you think? https://gist.github.com/flovilmart/bf64276507d685ae41f6#file-_pushstatus-L70

**Comment by aontas** [comment]
@flovilmart sounds good.
Should I assume that there will be an "audienceId" field around here: https://github.com/parse-community/parse-dashboard/blob/master/src/dashboard/Push/PushNew.react.js#L206 ?

**Comment by flovilmart** [comment]
Yep it's in the dashboard already: https://github.com/parse-community/parse-dashboard/blob/master/src/dashboard/Push/PushNew.react.js#L630 so we can basically transform the key audience_id to audienceId and go with it!

**Comment by aontas** [comment]
@flovilmart Have a look at those latest changes and see what you think.

File: spec/PushController.spec.js [hunk 2]
```diff
@@ -1063,7 +1059,11 @@ describe('PushController', () => {
         return query.find({ useMasterKey: true }).then(parseResults);
       });
     }).then(() => {
-      return pushController.sendPush(payload, where, config, auth)
+      var body = {data: {
+        alert: 'hello',
+        audience_id: audienceId
```
Comment (flovilmart) [inline spec/PushController.spec.js]: This would be into the body, not in data (topLevel) as data is sent to the user ;)

**Comment by flovilmart** [comment]
@aontas i’d Love to get that one in for 2.6.1 hopefully to release today!

File: src/Controllers/PushController.js [hunk 3]
```diff
@@ -55,6 +55,26 @@ export class PushController {
...
+        const audienceId = body.data.audience_id;
+        // Remove the audience_id from the push body.
+        delete body.data.audience_id;
+
+        var updateAudience = {
+          lastUsed: { __type: "Date", iso: new Date().toISOString() },
+          timesUsed: { __op: "Increment", "amount": 1 }
+        };
+        const audienceRestQuery = new RestQuery(config, master(config), '_Audience', {objectId: audienceId});
+        return audienceRestQuery.buildRestWhere().then(() => {
```
Comment (flovilmart) [inline src/Controllers/PushController.js]: Let’s not wait on that promise before sending the push? Because increments are atomic, we should be good

**Comment by aontas** [comment]
Based on https://github.com/parse-community/parse-dashboard/blob/master/src/dashboard/Push/PushNew.react.js#L198, the audience_id is a data member of payload. This means that it is within the data section, and making it top level will not work correctly.
(It's currently late in my timezone and I'll have to look at running up a dashboard and server together for testing tomorrow.)

**Comment by flovilmart** [comment]
Uhm ok... let’s see

**Comment by flovilmart** [comment]
@aontas actually, the audience_id is not sent to the server through the push call, we need to explicitly send it. 

As you can see [here](https://github.com/parse-community/parse-dashboard/blob/master/src/dashboard/Push/PushNew.react.js#L854) and [here](https://github.com/parse-community/parse-dashboard/blob/master/src/dashboard/Push/PushNew.react.js#L198) the audience_id is not part of the sent data. So I believe that would be better to pass it at the top level of the request.

File: src/Controllers/PushController.js [hunk 4]
```diff
@@ -56,16 +56,24 @@ export class PushController {
...
-        lastUsed: { __type: "Date", iso: new Date().toISOString() },
-        timesUsed: { __op: "Increment", "amount": 1 }
-      };
-      const audienceRestQuery = new RestQuery(config, master(config), '_Audience', {query: JSON.stringify(where)});
-      return audienceRestQuery.buildRestWhere().then(() => {
-        const write = new RestWrite(config, master(config), '_Audience', audienceRestQuery.restWhere, updateAudience);
-        write.runOptions.many = true;
-        return write.execute();
-      });
+      if (body.data && body.data.audience_id) {
```
Comment (flovilmart) [inline src/Controllers/PushController.js]: `if body.audience_id`

**Comment by aontas** [comment]
@flovilmart There you go. How does that look to you?

**Comment by aontas** [comment]
Oops, wrong comment button.

File: src/Controllers/PushController.js [hunk 5]
```diff
@@ -55,6 +55,24 @@ export class PushController {
...
+      if (body.audience_id) {
+        const audienceId = body.audience_id;
+
+        var updateAudience = {
+          lastUsed: { __type: "Date", iso: new Date().toISOString() },
+          timesUsed: { __op: "Increment", "amount": 1 }
+        };
+        const audienceRestQuery = new RestQuery(config, master(config), '_Audience', {objectId: audienceId});
+        audienceRestQuery.buildRestWhere().then(() => {
+          const write = new RestWrite(config, master(config), '_Audience', audienceRestQuery.restWhere, updateAudience);
```
Comment (flovilmart) [inline src/Controllers/PushController.js]: I believe we can use the basic RestWrite without building a restWhere query now as we only need objectId for it no?

**Comment by aontas** [comment]
... and done.

**Comment by flovilmart** [comment]
Those cases where less code is actually better :) Thanks for keeping it up! Let's update the dashboard now when we have a minute :)

**Pull request merged**